app/src/main/java/ch/sheremet/dasfoo/rover/android/client/MainActivity.java

```diff
@@ -32,116 +37,138 @@
     private EditText mPortEdit;
     private TextView mResultText;
 
+    private GrpcConnection mGrpcConnection;
+
+    private View.OnClickListener onClickListener = new View.OnClickListener() {
+        @Override
+        public void onClick(final View v) {
+            hideKeyboard();
+            switch (v.getId()) {
+                case R.id.move_forward_button:
+                    executeGrpcTask(new MovingRoverTask());
+                    break;
+                case R.id.info_button:
+                    executeGrpcTask(new GettingBoardInfoTask());
+                    break;
+                case R.id.read_encoders_button:
+                    executeGrpcTask(new EncodersReadingTask());
+                    break;
+                default:
+                    Log.v(TAG, "Button is not implemented yet.");
+                    break;
+            }
+        }
+    };
+
     @Override
-    protected void onCreate(Bundle savedInstanceState) {
+    protected void onCreate(final Bundle savedInstanceState) {
         super.onCreate(savedInstanceState);
         setContentView(R.layout.activity_main);
         mMoveForwardButton = (Button) findViewById(R.id.move_forward_button);
-        mMoveForwardButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                moveForward();
-            }
-        });
+        mMoveForwardButton.setOnClickListener(onClickListener);
         mInfoButton = (Button) findViewById(R.id.info_button);
-        mInfoButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                getInfo();
-            }
-        });
+        mInfoButton.setOnClickListener(onClickListener);
         mReadEncodersButton = (Button) findViewById(R.id.read_encoders_button);
-        mReadEncodersButton.setOnClickListener(new View.OnClickListener() {
-            @Override
-            public void onClick(View v) {
-                readEncoders();
-            }
-        });
+        mReadEncodersButton.setOnClickListener(onClickListener);
         mHostEdit = (EditText) findViewById(R.id.host_edit_text);
         mPortEdit = (EditText) findViewById(R.id.port_edit_text);
         mResultText = (TextView) findViewById(R.id.grpc_response_text);
     }
 
-    private void hideKeyboard(){
+    @Override
+    protected void onDestroy() {
+        super.onDestroy();
+        if (mGrpcConnection != null) {
+            mGrpcConnection.shutDownConnection();
```

Comment:
In case this method is called twice, might be better to assign `null` right here.e